libnmstate/nm/device.py

```diff
@@ -135,24 +135,33 @@ def _safe_activate_async(dev, connection_id):
 
 
 def _active_connection_callback(src_object, result, user_data):
-    mainloop, nmdev, cancellable = user_data
+    mainloop, nmdev, connection_id, cancellable = user_data
     mainloop.drop_cancellable(cancellable)
     try:
         nm_act_con = src_object.activate_connection_finish(result)
     except Exception as e:
+        if nmdev:
+            activation_type = "device"
+            activation_object = nmdev.get_iface()
+        elif connection_id:
+            activation_type = "connection_id"
+            activation_object = connection_id
+        else:
+            activation_type = activation_object = "unknown"
+
         if mainloop.is_action_canceled(e):
             logging.debug(
-                'Connection activation canceled on %s: error=%s',
-                nmdev.get_iface(), e)
+                'Connection activation canceled on %s %s: error=%s',
+                activation_type, activation_object, e)
         else:
             mainloop.quit(
-                'Connection activation failed on {}: error={}'.format(
-                    nmdev.get_iface(), e))
+                'Connection activation failed on {} {}: error={}'.format(
+                    activation_type, activation_object, e))
         return
 
     if nm_act_con is None:
-        mainloop.quit('Connection activation failed on %s: error=unknown' %
-                      nmdev.get_iface())
+        mainloop.quit('Connection activation failed on %s %s: error=unknown' %
+                      activation_type, activation_object)
```

Comment:
It cannot be done here, as these variables have not been defined yet.
With the helper func, you can just call it again.

Reply:
I ensured that the variables are defined now

Reply:
You just do it even if there is no need to do it (as this is relevant for errors only).

Reply:
changed